### ISSUE TITLE
BUG: Order of OS detection macros (for Windows) seems faulty

### ISSUE THREAD

**Issue by mahge** [numpy/numpy#24761]
### Describe the issue:

The order of these OS detection macros might make Win64 and MinGW unreachable.

https://github.com/numpy/numpy/blob/07db5becf57373b0962483ecaf3452b075183e1f/numpy/core/include/numpy/npy_os.h#L22-L27

According to Microsoft [documentation](https://learn.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=msvc-170) the macros are defined as follows:

>> _WIN32 Defined as 1 when the compilation target is 32-bit ARM, 64-bit ARM, x86, or x64. Otherwise, undefined.
>> _WIN64 Defined as 1 when the compilation target is 64-bit ARM or x64. Otherwise, undefined.

Essentially `_WIN32` is defined on 32 bit as well as 64 bit setups. `_WIN64` is defined on 64 bit Windows only. This means the first check 
```cpp
#elif defined(_WIN32) || defined(__WIN32__) || defined(WIN32) 
     #define NPY_OS_WIN32 
```

will succeed and the numpy OS will be defined as `NPY_OS_WIN32` on 64-bit ARM or x64 Windows.

On top of that, MinGW compilers follow the same rule. So `_WIN32` is defined, say on 32 bit as well as 64 bit `MinGW64` `gcc/clang` while `_WIN64` is defined for 64 bit compilations only. 

Although I can not find a definite documentation for the MinGW64 compiler macros, as far as I can tell this is the state of these macros:

    64 bit MinGW compilers define `_MINGW32__`, `__MINGW64__`, `_WIN32`, and `_WIN64`.
    32 bit MinGW compilers define `__MINGW32__`, and `_WIN32`.
    64 bit MSVC compilation defines `_WIN32` and  `_WIN64`.
    32 bit MSVC compilation defines `_WIN32`.

Therefore, on 32 bit or 64 bit Windows or MinGW, the OS check will always say that the OS is `NPY_OS_WIN32`. This might be fine for most use cases (especially since the macros do not seem to be used in many places) but it is also not what is intended. 

---------------------------------------------------

With MinGW's 64-bit UCRT64 gcc, for example, this causes compilation failure because the OS is defined as `NPY_OS_WIN32` and this causes this check    

```cpp
#ifndef F2PY_THREAD_LOCAL_DECL
#if defined(_MSC_VER)
#define F2PY_THREAD_LOCAL_DECL __declspec(thread)
#elif defined(NPY_OS_MINGW)
#define F2PY_THREAD_LOCAL_DECL __thread
#elif defined(__STDC_VERSION__) \
      && (__STDC_VERSION__ >= 201112L) \
      && !defined(__STDC_NO_THREADS__) \
      && (!defined(__GLIBC__) || __GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ > 12)) \
      && !defined(NPY_OS_OPENBSD) && !defined(NPY_OS_HAIKU)
/* __STDC_NO_THREADS__ was first defined in a maintenance release of glibc 2.12,
   see https://lists.gnu.org/archive/html/commit-hurd/2012-07/msg00180.html,
   so `!defined(__STDC_NO_THREADS__)` may give false positive for the existence
   of `threads.h` when using an older release of glibc 2.12
   See gh-19437 for details on OpenBSD */
#include <threads.h>
#define F2PY_THREAD_LOCAL_DECL thread_local
#elif defined(__GNUC__) \
      && (__GNUC__ > 4 || (__GNUC__ == 4 && (__GNUC_MINOR__ >= 4)))
#define F2PY_THREAD_LOCAL_DECL __thread
#endif
#endif
```

**in the generated files when using f2py** to assume `threads.h` is available. Compilation fails because it is not available.
  

### Reproduce the code example:

```python
Unfortunately I do not have a simple example to reproduce this. I am also not very familiar with numpy or f2py as much as I would like. I came across it while fixing things in a bigger project. 

On the other hand it should be easier to reproduce for someone who know what they are doing and have access to MYSY's UCRT64 environment. Even if that is not possible one can possible figure out what the results might be by looking at the relevant section in npy_os.h and the reported predefined macro status.
```

### Error message:

```sh
msys/tmp/tmp1rhc293t/src.mingw_x86_64_ucrt-3.11/curvif_simplifiedmodule.c:87:10: fatal error: threads.h: No such file or directory
   87 | #include <threads.h>
      |          ^~~~~~~~~~~
compilation terminated.
```

### Runtime information:

```sh
$> f2py

Version:     1.25.2
numpy Version: 1.25.2
Requires:    Python 3.5 or higher.
License:     NumPy license (see LICENSE.txt in the NumPy source code)
Copyright 1999 - 2011 Pearu Peterson all rights reserved.
https://web.archive.org/web/20140822061353/http://cens.ioc.ee/projects/f2py2e
```

**Comment by mattip** [comment]
I am a bit confused why this has not come up before. I see the snippet you reproduced with the `include <threads.h>` comes from f2py [here](https://github.com/numpy/numpy/blob/d0c6cdf8da4db4a2240dbbc6baa64bc01f27e05b/numpy/f2py/cfuncs.py#L543C31-L543C31) in f2py, is this not compiled into tests? Or maybe the rtools mingw we use for fortran compilation does not use the right runtime to hit the bug?

@HaoZeke as the last one to touch that file in #20884 do you have any insights?

**Comment by mahge** [comment]
> I am a bit confused why this has not come up before. 

My guess is that some of the other conditions in the check

```cpp
#elif defined(__STDC_VERSION__) \
      && (__STDC_VERSION__ >= 201112L) \
      && !defined(__STDC_NO_THREADS__) \
      && (!defined(__GLIBC__) || __GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ > 12)) \
      && !defined(NPY_OS_OPENBSD) && !defined(NPY_OS_HAIKU)
/* __STDC_NO_THREADS__ was first defined in a maintenance release of glibc 2.12,
   see https://lists.gnu.org/archive/html/commit-hurd/2012-07/msg00180.html,
   so `!defined(__STDC_NO_THREADS__)` may give false positive for the existence
   of `threads.h` when using an older release of glibc 2.12
   See gh-19437 for details on OpenBSD */
#include <threads.h>
#define F2PY_THREAD_LOCAL_DECL thread_local
```

were not satisfied (for most this is probably due to requesting `__STDC_VERSION__ >= 201112L`). This meant the check probably got satisfied by the next check

```cpp
#elif defined(__GNUC__) \
      && (__GNUC__ > 4 || (__GNUC__ == 4 && (__GNUC_MINOR__ >= 4)))
#define F2PY_THREAD_LOCAL_DECL __thread
```

which, coincidentally, has the same effect as NPY_OS_MINGW being defined anyway. Considering this last check was for `gcc > 4.4`, it is likely it was satisfied by most platforms by the time the change was made. 

In MinGW UCRT64¨s gcc-13 case, I can see we have `__STDC_VERSION__ >= 201112L` satisfied. Maybe that is the reason why it popped up now.  I am guessing the explicit check for MinGW environments was made to address this issue.

**Comment by HaoZeke** [comment]
> I am a bit confused why this has not come up before. I see the snippet you reproduced with the `include <threads.h>` comes from f2py [here](https://github.com/numpy/numpy/blob/d0c6cdf8da4db4a2240dbbc6baa64bc01f27e05b/numpy/f2py/cfuncs.py#L543C31-L543C31) in f2py, is this not compiled into tests? Or maybe the rtools mingw we use for fortran compilation does not use the right runtime to hit the bug?
> 
> @HaoZeke as the last one to touch that file in #20884 do you have any insights?

As I recall, the changes were to go from hand-crafted macros to the ones used in the rest of NumPy, and wouldn't have handled / caught this bug.

**Comment by HaoZeke** [comment]
FWIW a lot of the macro logic in `f2py` can be simplified with `meson` configuration files once that becomes the default, but that's a separate issue :)

**Comment by carlkl** [comment]
predefined macros, look here: https://github.com/cpredef/predef

**Comment by carlkl** [comment]
Sorry for my ignorance, What is the purpose of `NPY_OS_MINGW` within numpy?

Do we also want to differentiate between `mingw32` and `mingw-w64` toolchain? The `mingw32` toolchain is the now obsolet precedessor of `mingw-w64` and should issue a warning IMHO.

Do we want to differentiate between 32-bit and 64-bit version of the `mingw-w64` toolchain?

Do we need to check the gcc version and/or the version of the mingw-w64 headers used, to ensure a minimum of features / issue a warning?

**Comment by mahge** [comment]
> Do we also want to differentiate between `mingw32` and `mingw-w64` toolchain? The `mingw32` toolchain is the now obsolet precedessor of `mingw-w64` and should issue a warning IMHO.
> 
> Do we want to differentiate between 32-bit and 64-bit version of the `mingw-w64` toolchain?

I think it is more of differentiating `Windows` (as in MSVC) vs `MinGW-w64` compilers rather than detecting the architecture. At least that seems to be the intention. Of course it does not hurt if the detection also figures out the architecture when possible. The purpose of `NPY_OS_MINGW` seems to be identifying that we are on Windows but using one of the `MinGW-w64` compilers. The definition OS/compiler gets a bit mixed up when refereeing to MSYS/MinGW environments. 

Of course these OS/platform/compiler detections are never really exhaustive and get fixed/built-up overtime. It all depends on if numpy uses some features of a specific combination of these OS/platform/compiler variants.

**Comment by mahge** [comment]
> FWIW a lot of the macro logic in f2py can be simplified with meson configuration files once that becomes the default, but that's a separate issue :)

I am not sure if it helps that much in this specific case. The configuration tools like Mason and CMake will give you the macros needed sure. However that will not save you from the fact that _WIN32 will be defined for 64 bit Windows compilations with MSVC or MinGW-w64 compilers. It is by design. We just have to pick and choose how we use those macros. 

CMake used to have a module called [WriteCompilerDetectionHeader](https://cmake.org/cmake/help/latest/module/WriteCompilerDetectionHeader.html) that tried to help in detecting different compilers by generating a header file that can be included by the project. Unfortunately, they have now deprecated it. Maybe Meson does a good job at it. 

The CMake generated compiler detection header will look something like this. It is probably an overkill but can also be a good starting point if you want to be as exhaustive as possible.

<details>
<summary>numpy_compiler_detection.h</summary>

```cpp



// This is a generated file. Do not edit!

#ifndef NUMPY_COMPILER_DETECTION_H
#define NUMPY_COMPILER_DETECTION_H

#ifdef __cplusplus
#define NUMPY_COMPILER_IS_Comeau 0
#define NUMPY_COMPILER_IS_Intel 0
#define NUMPY_COMPILER_IS_PathScale 0
#define NUMPY_COMPILER_IS_Embarcadero 0
#define NUMPY_COMPILER_IS_Borland 0
#define NUMPY_COMPILER_IS_Watcom 0
#define NUMPY_COMPILER_IS_OpenWatcom 0
#define NUMPY_COMPILER_IS_SunPro 0
#define NUMPY_COMPILER_IS_HP 0
#define NUMPY_COMPILER_IS_Compaq 0
#define NUMPY_COMPILER_IS_zOS 0
#define NUMPY_COMPILER_IS_XLClang 0
#define NUMPY_COMPILER_IS_XL 0
#define NUMPY_COMPILER_IS_VisualAge 0
#define NUMPY_COMPILER_IS_PGI 0
#define NUMPY_COMPILER_IS_Cray 0
#define NUMPY_COMPILER_IS_TI 0
#define NUMPY_COMPILER_IS_Fujitsu 0
#define NUMPY_COMPILER_IS_GHS 0
#define NUMPY_COMPILER_IS_SCO 0
#define NUMPY_COMPILER_IS_ARMCC 0
#define NUMPY_COMPILER_IS_AppleClang 0
#define NUMPY_COMPILER_IS_ARMClang 0
#define NUMPY_COMPILER_IS_Clang 0
#define NUMPY_COMPILER_IS_GNU 0
#define NUMPY_COMPILER_IS_MSVC 0
#define NUMPY_COMPILER_IS_ADSP 0
#define NUMPY_COMPILER_IS_IAR 0
#define NUMPY_COMPILER_IS_MIPSpro 0

#if defined(__COMO__)
#undef NUMPY_COMPILER_IS_Comeau
#define NUMPY_COMPILER_IS_Comeau 1

#elif defined(__INTEL_COMPILER) || defined(__ICC)
#undef NUMPY_COMPILER_IS_Intel
#define NUMPY_COMPILER_IS_Intel 1

#elif defined(__PATHCC__)
#undef NUMPY_COMPILER_IS_PathScale
#define NUMPY_COMPILER_IS_PathScale 1

#elif defined(__BORLANDC__) && defined(__CODEGEARC_VERSION__)
#undef NUMPY_COMPILER_IS_Embarcadero
#define NUMPY_COMPILER_IS_Embarcadero 1

#elif defined(__BORLANDC__)
#undef NUMPY_COMPILER_IS_Borland
#define NUMPY_COMPILER_IS_Borland 1

#elif defined(__WATCNUMPY__) && __WATCNUMPY__ < 1200
#undef NUMPY_COMPILER_IS_Watcom
#define NUMPY_COMPILER_IS_Watcom 1

#elif defined(__WATCNUMPY__)
#undef NUMPY_COMPILER_IS_OpenWatcom
#define NUMPY_COMPILER_IS_OpenWatcom 1

#elif defined(__SUNPRO_CC)
#undef NUMPY_COMPILER_IS_SunPro
#define NUMPY_COMPILER_IS_SunPro 1

#elif defined(__HP_aCC)
#undef NUMPY_COMPILER_IS_HP
#define NUMPY_COMPILER_IS_HP 1

#elif defined(__DECCXX)
#undef NUMPY_COMPILER_IS_Compaq
#define NUMPY_COMPILER_IS_Compaq 1

#elif defined(__IBMCPP__) && defined(__COMPILER_VER__)
#undef NUMPY_COMPILER_IS_zOS
#define NUMPY_COMPILER_IS_zOS 1

#elif defined(__ibmxl__) && defined(__clang__)
#undef NUMPY_COMPILER_IS_XLClang
#define NUMPY_COMPILER_IS_XLClang 1

#elif defined(__IBMCPP__) && !defined(__COMPILER_VER__) && __IBMCPP__ >= 800
#undef NUMPY_COMPILER_IS_XL
#define NUMPY_COMPILER_IS_XL 1

#elif defined(__IBMCPP__) && !defined(__COMPILER_VER__) && __IBMCPP__ < 800
#undef NUMPY_COMPILER_IS_VisualAge
#define NUMPY_COMPILER_IS_VisualAge 1

#elif defined(__PGI)
#undef NUMPY_COMPILER_IS_PGI
#define NUMPY_COMPILER_IS_PGI 1

#elif defined(_CRAYC)
#undef NUMPY_COMPILER_IS_Cray
#define NUMPY_COMPILER_IS_Cray 1

#elif defined(__TI_COMPILER_VERSION__)
#undef NUMPY_COMPILER_IS_TI
#define NUMPY_COMPILER_IS_TI 1

#elif defined(__FUJITSU) || defined(__FCC_VERSION) || defined(__fcc_version)
#undef NUMPY_COMPILER_IS_Fujitsu
#define NUMPY_COMPILER_IS_Fujitsu 1

#elif defined(__ghs__)
#undef NUMPY_COMPILER_IS_GHS
#define NUMPY_COMPILER_IS_GHS 1

#elif defined(__SCO_VERSION__)
#undef NUMPY_COMPILER_IS_SCO
#define NUMPY_COMPILER_IS_SCO 1

#elif defined(__ARMCC_VERSION) && !defined(__clang__)
#undef NUMPY_COMPILER_IS_ARMCC
#define NUMPY_COMPILER_IS_ARMCC 1

#elif defined(__clang__) && defined(__apple_build_version__)
#undef NUMPY_COMPILER_IS_AppleClang
#define NUMPY_COMPILER_IS_AppleClang 1

#elif defined(__clang__) && defined(__ARMCOMPILER_VERSION)
#undef NUMPY_COMPILER_IS_ARMClang
#define NUMPY_COMPILER_IS_ARMClang 1

#elif defined(__clang__)
#undef NUMPY_COMPILER_IS_Clang
#define NUMPY_COMPILER_IS_Clang 1

#elif defined(__GNUC__) || defined(__GNUG__)
#undef NUMPY_COMPILER_IS_GNU
#define NUMPY_COMPILER_IS_GNU 1

#elif defined(_MSC_VER)
#undef NUMPY_COMPILER_IS_MSVC
#define NUMPY_COMPILER_IS_MSVC 1

#elif defined(__VISUALDSPVERSION__) || defined(__ADSPBLACKFIN__) || defined(__ADSPTS__) || defined(__ADSP21000__)
#undef NUMPY_COMPILER_IS_ADSP
#define NUMPY_COMPILER_IS_ADSP 1

#elif defined(__IAR_SYSTEMS_ICC__) || defined(__IAR_SYSTEMS_ICC)
#undef NUMPY_COMPILER_IS_IAR
#define NUMPY_COMPILER_IS_IAR 1

#endif

#if NUMPY_COMPILER_IS_GNU

#if !((__GNUC__ * 100 + __GNUC_MINOR__) >= 404)
  #error Unsupported compiler version
#endif

#if defined(__GNUC__)
  #define NUMPY_COMPILER_VERSION_MAJOR (__GNUC__)
#else
  #define NUMPY_COMPILER_VERSION_MAJOR (__GNUG__)
#endif
  #if defined(__GNUC_MINOR__)
    #define NUMPY_COMPILER_VERSION_MINOR (__GNUC_MINOR__)
  #endif
  #if defined(__GNUC_PATCHLEVEL__)
  #define NUMPY_COMPILER_VERSION_PATCH (__GNUC_PATCHLEVEL__)
  #endif

#elif NUMPY_COMPILER_IS_Clang

  #if !(((__clang_major__ * 100) + __clang_minor__) >= 301)
    #error Unsupported compiler version
  #endif

  #define NUMPY_COMPILER_VERSION_MAJOR (__clang_major__)
  #define NUMPY_COMPILER_VERSION_MINOR (__clang_minor__)
  #define NUMPY_COMPILER_VERSION_PATCH (__clang_patchlevel__)

  #if defined(_MSC_VER)
    /* _MSC_VER = VVRR */
    #define NUMPY_SIMULATE_VERSION_MAJOR (_MSC_VER / 100)
    #define NUMPY_SIMULATE_VERSION_MINOR (_MSC_VER % 100)
  #endif

#elif NUMPY_COMPILER_IS_MSVC

  #if !(_MSC_VER >= 1600)
    #error Unsupported compiler version
  #endif

  /* _MSC_VER = VVRR */
  #define NUMPY_COMPILER_VERSION_MAJOR (_MSC_VER / 100)
  #define NUMPY_COMPILER_VERSION_MINOR (_MSC_VER % 100)
  #if defined(_MSC_FULL_VER)
    #if _MSC_VER >= 1400
      /* _MSC_FULL_VER = VVRRPPPPP */
      #define NUMPY_COMPILER_VERSION_PATCH (_MSC_FULL_VER % 100000)
    #else
      /* _MSC_FULL_VER = VVRRPPPP */
      #define NUMPY_COMPILER_VERSION_PATCH (_MSC_FULL_VER % 10000)
    #endif
  #endif
  #if defined(_MSC_BUILD)
    #define NUMPY_COMPILER_VERSION_TWEAK (_MSC_BUILD)
  #endif

#else
  #error Unsupported compiler
#endif

#endif

#endif

```
</details>

**Comment by DimitriPapadopoulos** [comment]
Please retire **`WIN32`** and **`WIN64`** in the process, keeping `_WIN32` and `_WIN64` of course. As far as I can understand and according to [QM Bites: Understand Windows OS Identification Preprocessor Macros](https://accu.org/journals/overload/24/132/wilson_2223/):
* **`_WIN32`** and **`_WIN64`** are defined **by the compiler**,
* **`WIN32`** and **`WIN64`** are defined **by the user**, to indicate whatever the user chooses them to indicate. They mean 32-bit and 64-bit Windows compilation by convention only.
